### PR TITLE
Feature/preact

### PR DESCRIPTION
**Pull request by lucianlature**
The wait is over. `Preact` is now powering `React-Universally`, both client and server side.

Implementation:
- for the client, `webpack` aliases are configured, so that `preact` replaces `react`.
- for the server, things are a little bit different, since `webpack` is skipped when requiring modules, so I had to make use of `module-alias` in order to operate the replacement in `server` mode.
- probably I will have to come up with some benchmarks to prove the differences in performance when using one or the other.

Please take a look over it and wait few days before merging in order to gain some maturity :)

### CODE REVIEW

**Comment by ctrlplusb** [comment]
Awesome!! Thanks @lucianlature :)

File: config/index.js [hunk 1]
```diff
@@ -237,6 +244,33 @@ const config = {
           'react-router',
         ],
 
+        // Which libaries should be excluded when interpreting the package.json
```
Comment (ctrlplusb) [inline config/index.js]: Woops, this was dropped previously.  I think the merge from master went awry.

**Comment by lucianlature** [comment]
@ctrlplusb Regarding `Inferno`, it does not play nice with `Miss` component of `react-router`. Other than that, I have a branch that's using `Inferno` instead of `React`. Should I push that one as well or wait until all the issues are being fixed and ready for production?

File: config/index.js (hunk 1 above)
Comment (lucianlature) [inline config/index.js]: No worries, I will merge again and fix it

File: src/client/index.js [hunk 2]
```diff
@@ -4,9 +4,14 @@ import React from 'react';
...
 import { BrowserRouter } from 'react-router';
 import { CodeSplitProvider, rehydrateState } from 'code-split-component';
-import ReactHotLoader from './components/ReactHotLoader';
+// import ReactHotLoader from './components/ReactHotLoader';
 import DemoApp from '../shared/components/DemoApp';
 
+// Enable devtools. You can reduce the size of your app by only including this
+// module in development builds. eg. In Webpack, wrap this with an `if (module.hot) {...}`
+// check.
+require('preact/devtools');
```
Comment (ctrlplusb) [inline src/client/index.js]: We could move this to the webpack config factory, adding it to the entries for the client dev bundle.

Comment (lucianlature) [inline src/client/index.js]: :+1:

File: src/server/index.js [hunk 3]
```diff
@@ -1,5 +1,9 @@
 /* eslint-disable no-console */
 
+// This grants us source map support, which combined with our webpack source
+// maps will give us nice stack traces.
+import 'source-map-support/register';
```
Comment (ctrlplusb) [inline src/server/index.js]: Another merge issue, we moved this to the webpack config factory.

File: tools/development/index-await.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import path, { resolve as pathResolve } from 'path';
```
Comment (ctrlplusb) [inline tools/development/index-await.js]: Woah. :) What is this? :)

File: tools/webpack/configFactory.js [hunk 5]
```diff
@@ -139,7 +139,7 @@ export default function webpackConfigFactory(buildOptions) {
         // bundles.
         ifNode('source-map-support/register'),
         // Required to support hot reloading of our client.
-        ifDevClient('react-hot-loader/patch'),
+        // ifDevClient('react-hot-loader/patch'),
```
Comment (ctrlplusb) [inline tools/webpack/configFactory.js]: I would remove this completely.

Comment (lucianlature) [inline tools/webpack/configFactory.js]: Yep, the `HMR` still works without it

File: tools/webpack/configFactory.js [hunk 6]
```diff
@@ -565,6 +566,8 @@ export default function webpackConfigFactory(buildOptions) {
               path.resolve(appRootDir.get(), srcPath),
             ),
             ifProdClient(path.resolve(appRootDir.get(), 'src/html')),
+            // preact-compat has some es6 requirements. :)
+            // path.resolve(appRootDir.get(), './node_modules/preact-compat'),
```
Comment (ctrlplusb) [inline tools/webpack/configFactory.js]: Is this needed?

Comment (lucianlature) [inline tools/webpack/configFactory.js]: Nope, should be removed.

File: tools/webpack/configFactory.js (hunk 5 above)
Comment (ctrlplusb) [inline tools/webpack/configFactory.js]: Also, do you mind doing a yarn remove on the dependency.

File: src/client/index.js [hunk 7]
```diff
@@ -22,15 +27,13 @@ function renderApp(TheApp) {
   // @see https://github.com/ctrlplusb/code-split-component
   rehydrateState().then(codeSplitState =>
     render(
-      <ReactHotLoader>
```
Comment (ctrlplusb) [inline src/client/index.js]: We should delete the ReactHotLoader component too.

**Comment by lucianlature** [comment]
@ctrlplusb Should I use `next` or `master` for merging?

File: src/client/index.js [hunk 8]
```diff
@@ -4,9 +4,14 @@ import React from 'react';
 import { render } from 'react-dom';
 import { BrowserRouter } from 'react-router';
 import { CodeSplitProvider, rehydrateState } from 'code-split-component';
-import ReactHotLoader from './components/ReactHotLoader';
+// import ReactHotLoader from './components/ReactHotLoader';
```
Comment (ctrlplusb) [inline src/client/index.js]: I would remove this completely.

**Comment by ctrlplusb** [comment]
Use `master` at the moment.

File: tools/development/index-await.js (hunk 4 above)
Comment (ctrlplusb) [inline tools/development/index-await.js]: I mean the entire index-await.js file :)

File: tools/webpack/configFactory.js (hunk 5 above)
Comment (lucianlature) [inline tools/webpack/configFactory.js]: Done already ;)

File: tools/development/index-await.js (hunk 4 above)
Comment (lucianlature) [inline tools/development/index-await.js]: Don't know, I'll get rid of it

**Comment by ctrlplusb** [comment]
Nice looking good! :)

**Comment by ctrlplusb** [comment]
🎉

**Pull request merged**